### PR TITLE
Add internal solution for Searching for Soulmates (usaco-1182)

### LINKED ISSUE #3757: Searching For Soulmates

### PR DESCRIPTION
**Pull request by ChuangSheep**
Closes #3757 

_If any of the below don't apply to this Pull Request, mark the checkbox as done._

- [x] I have tested my code.
- [x] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [x] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions).
  - I understand that if it is clear that I have not attempted to follow these conventions, my PR will be closed.
  - If changes are requested, I will re-request a review after addressing them.

### CODE REVIEW

File: solutions/silver/usaco-1182.mdx [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+author: Chongtian Ma, Chuyang Wang
+---
+
+[Official Analysis (Java)](http://www.usaco.org/current/data/sol_prob1_silver_jan22.html)
+
+
+## Explanation
+
+To transform $a$ into $b$, we proceed greedily. If $a > b$, it can be shown that
+it is never optimal to add more than once consecutively before dividing $a$ by
```
Comment (SansPapyrus683) [inline solutions/silver/usaco-1182.mdx:14]: > it can be shown

can you at least give some intuition

File: solutions/silver/usaco-1182.mdx [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+		 * Otherwise, simulate multiplication of a by dividing b.
+		 * Alternatively, if it's better, add up to b directly.
+		 */
+		ll is_odd = b % 2;
+		return min(b - a, 1 + is_odd + solve(a, b / 2));
+	}
+}
+
+int main() {
+	int N;
```
Comment (SansPapyrus683) [inline solutions/silver/usaco-1182.mdx:54]: can you use lowercase

File: solutions/silver/usaco-1182.mdx [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+```
+
+</CPPSection>
+<JavaSection>
+
+```java
+import java.io.*;
+import java.util.*;
+
+public class solution {
```
Comment (SansPapyrus683) [inline solutions/silver/usaco-1182.mdx:71]:
```suggestion
public class Soulmates {
```

File: solutions/silver/usaco-1182.mdx [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+			/*
+			 * Divide a greedily until a <= b, add 1 in case a is odd to enable
+			 * division.
+			 */
+			long isOdd = a % 2;
+			return 1 + isOdd + solve((a + isOdd) / 2, b);
+		} else {
+			/*
+			 * Otherwise, simulate multiplication of a by dividing b.
+			 * Alternatively, if it's better, add up to b directly.
```
Comment (SansPapyrus683) [inline solutions/silver/usaco-1182.mdx:85]: why do we have to *simulate* multiplication of a?

File: solutions/silver/usaco-1182.mdx [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+			 */
+			long isOdd = b % 2;
+			return Math.min(b - a, 1 + isOdd + solve(a, b / 2));
+		}
+	}
+
+	public static void main(String[] args) throws IOException {
+		BufferedReader br =
+		    new BufferedReader(new InputStreamReader(System.in));
+		int N = Integer.parseInt(br.readLine());
```
Comment (SansPapyrus683) [inline solutions/silver/usaco-1182.mdx:95]: yeah here too

File: solutions/silver/usaco-1182.mdx [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+		Otherwise, simulate multiplication of a by dividing b.
+		Alternatively, if it's better, add up to b directly.
+		"""
+		is_odd = b % 2
+		return min(b - a, 1 + is_odd + solve(a, b // 2))
+
+
+N = int(input())
+
+for i in range(N):
```
Comment (SansPapyrus683) [inline solutions/silver/usaco-1182.mdx:133]:
```suggestion
for _ in range(int(input())):
```

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

File: solutions/silver/usaco-1182.mdx (hunk 4 above)
Comment (ChuangSheep) [reply solutions/silver/usaco-1182.mdx:85]: ~~cause i copied envy's code~~
cause we're only allowed to modify a. What we do on b can be seen as operations on a in reversed order.

File: solutions/silver/usaco-1182.mdx (hunk 2 above)
Comment (ChuangSheep) [reply solutions/silver/usaco-1182.mdx:54]: it's uppercase in the problem statement

Comment (SansPapyrus683) [reply solutions/silver/usaco-1182.mdx:58]: yeah but c++ is snake_case so

File: solutions/silver/usaco-1182.mdx (hunk 4 above)
Comment (SansPapyrus683) [reply solutions/silver/usaco-1182.mdx:89]: can you do that op on a to make it more explicit

**Comment by SansPapyrus683** [comment]
bro i legit got so scared when i saw your commit name

File: solutions/silver/usaco-1182.mdx (hunk 4 above)
Comment (ChuangSheep) [reply solutions/silver/usaco-1182.mdx:89]: tbh im not sure if we can do it directly on `a` without making the recursion very ugly

File: solutions/silver/usaco-1182.mdx (hunk 2 above)
Comment (ChuangSheep) [reply solutions/silver/usaco-1182.mdx]: tbh i don't really agree with this but ok

**Review by envyaims — APPROVED** [review]
great, but since you mention simulate a lot you could sort of explain the complexity as well

File: solutions/silver/usaco-1182.mdx (hunk 4 above)
Comment (SansPapyrus683) [reply solutions/silver/usaco-1182.mdx:89]: but looking at it we should be able to do this in just 1 step if b is even & 2 is b is odd?

**Comment by ChuangSheep** [comment]
alright i think the word "simulate" is not a good choice. Ima gonna rewrite this part

File: solutions/silver/usaco-1182.mdx (hunk 4 above)
Comment (ChuangSheep) [reply solutions/silver/usaco-1182.mdx:89]: can u elaborate, esp. how to determine how many times u add at the beginning if u work on a?

Comment (SansPapyrus683) [reply solutions/silver/usaco-1182.mdx:89]: wait what
is this question meant for me

File: solutions/silver/usaco-1182.mdx [hunk 7]
```diff
@@ -16,10 +16,12 @@ it is never optimal to add more than once consecutively before dividing $a$ by
 to three operations in sum, we can instead first divide and then just add one,
 which has the same effect but requires only two operations.
 
-Then, if $a < b$, we can simulate our strategy above by allowing multiplication,
-division, and subtraction by one on $b$. We therefore either divide $b$ by two
-(and first subtract by one if $b$ is odd) or subtract $b$ $b - a$ times to
-directly reach $a$.
+Then, if $a < b$, as the optimality of the strategy above is already shown,
+let's reuse it and work on $b$ by allowing multiplication, division, and
+subtraction by one. We therefore either divide $b$ by two (and first subtract by
```
Comment (SansPapyrus683) [inline solutions/silver/usaco-1182.mdx:21]: doesn't the problem only allow you to work on a

Comment (ChuangSheep) [reply solutions/silver/usaco-1182.mdx:21]: as stated, these two ops are equivalent
meaning what we do on b is exactly what we can do on a

File: solutions/silver/usaco-1182.mdx (hunk 4 above)
Comment (ChuangSheep) [reply solutions/silver/usaco-1182.mdx]: > we should be able to do this in just 1 step if b is even & 2 is b is odd?

emm what

**Comment by SansPapyrus683** [comment]
god i am bad at reading comprehension

**Pull request merged**